assets/js/blocks/classic-template/index.tsx

```diff
@@ -99,51 +105,135 @@ const Edit = ( { clientId, attributes }: Props ) => {
+			multiple: false,
+			reusable: false,
+			inserter,
 		},
-	},
-	attributes: {
-		/**
-		 * Template attribute is used to determine which core PHP template gets rendered.
-		 */
-		template: {
-			type: 'string',
-			default: 'any',
+		example: {
+			attributes: {
+				isPreview: true,
+			},
 		},
-		align: {
-			type: 'string',
-			default: 'wide',
+		attributes: {
+			/**
+			 * Template attribute is used to determine which core PHP template gets rendered.
+			 */
+			template: {
+				type: 'string',
+				default: 'any',
+			},
+			align: {
+				type: 'string',
+				default: 'wide',
+			},
 		},
-	},
-	edit: Edit,
-	save: () => null,
-} );
+		edit: ( { attributes, clientId } ) => {
+			return (
+				<Edit
+					attributes={ {
+						...attributes,
+						template: template ?? ( attributes.template as string ),
+					} }
+					clientId={ clientId }
+				/>
+			);
+		},
+		save: () => null,
+	} );
+};
+
+const isClassicTemplateBlockRegisteredWithAnotherTitle = (
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	block: Block< any > | undefined,
+	parsedTemplate: string | undefined
+) => parsedTemplate && block?.title !== TEMPLATES[ parsedTemplate ].title;
+
+const hasTemplateSupportForClassicTemplateBlock = (
+	parsedTemplate: string | undefined
+) => parsedTemplate && templates.includes( parsedTemplate );
+
+// @todo Refactor when there will be possible to show a block according on a template/post with a Gutenberg API. https://github.com/WordPress/gutenberg/pull/41718
+
+let currentTemplateId: string | undefined;
+
+if ( isExperimentalBuild() ) {
+	subscribe( () => {
+		const previousTemplateId = currentTemplateId;
+		const store = select( 'core/edit-site' );
+		currentTemplateId = store?.getEditedPostId() as string | undefined;
+
+		if ( previousTemplateId === currentTemplateId ) {
+			return;
+		}
+
+		const parsedTemplate = currentTemplateId?.split( '//' )[ 1 ];
+		const block = getBlockType( BLOCK_SLUG );
+
+		if (
+			block !== undefined &&
+			( ! hasTemplateSupportForClassicTemplateBlock( parsedTemplate ) ||
+				isClassicTemplateBlockRegisteredWithAnotherTitle(
+					block,
+					parsedTemplate
+				) )
+		) {
+			unregisterBlockType( BLOCK_SLUG );
+			currentTemplateId = undefined;
+			return;
+		}
+
+		if (
+			block === undefined &&
+			hasTemplateSupportForClassicTemplateBlock( parsedTemplate )
+		) {
+			/**
+			 * The 'WooCommerce Legacy Template' block was renamed to 'WooCommerce Classic Template', however, the internal block
+			 * name 'woocommerce/legacy-template' needs to remain the same. Otherwise, it would result in a corrupt block when
+			 * loaded for users who have customized templates using the legacy-template (since the internal block name is
+			 * stored in the database).
+			 *
+			 * See https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5861 for more context
+			 */
+			registerClassicTemplateBlock( {
+				template: parsedTemplate ?? 'WooCommerce Classic Template',
```

Comment:
Because of the condition in line 219, Isn't the `parsedTemplate` variable always truly at this point? By the way, 'WooCommerce Classic Template' isn't the correct value for template, it should be the template ID.